Refactor: Require scriptPubKey to get wallet SigningProvider

achow101 opened the pull request:
Easier to review ignoring whitespace:

    git log -p -n1 -w

This commit does not change behavior. It passes new CScript arguments to
signing functions, but the arguments aren't currently used.

Split from #17261

achow101 committed:
Refactor: Require scriptPubKey to get wallet SigningProvider

Easier to review ignoring whitespace:

    git log -p -n1 -w

This commit does not change behavior. It passes new CScript arguments to
signing functions, but the arguments aren't currently used.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3281,7 +3290,23 @@ UniValue signrawtransactionwithwallet(const JSONRPCRequest& request)
+
+    UniValue result;
+    for (const SigningProvider* provider : providers) {
+        result = SignTransaction(mtx, provider, coins, request.params[2]);
```
There shouldn't ever be a SigningProvider that doesn't match.

ryanofsky replied:
In commit "Refactor: Require scriptPubKey to get wallet SigningProvider" (f7510c144b2262e1ee114eaa740b889c24ec7abb)

According to RPC documentation result has three fields: "hex" "complete" and "errors". I guess for "hex" and "complete" fields it makes sense to take the last values, but errors field should probably accumulate? In any case, this code is not very future proof in case new fields are added. Would suggest changing SignTransaction to take a `UniValue& result` input/output parameter, so it can be responsible for updating the result at each step, and this code isn't just blindly overwriting previous results.

achow101 replied:
I've added a commit which changes `SignTransaction` to `void` and has a `result` parameter to accumulate the errors array.

achow101 commented on src/wallet/psbtwallet.cpp:
```diff
@@ -39,12 +39,33 @@ TransactionError FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& ps
+        // Get the scriptPubKey to know which SigningProvider to use
+        CScript script;
+        if (input.non_witness_utxo) {
+            script = input.non_witness_utxo->vout[txin.prevout.n].scriptPubKey;
```
Done

achow101 commented on src/wallet/psbtwallet.cpp:
```diff
@@ -39,12 +39,33 @@ TransactionError FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& ps
+        } else if (!input.witness_utxo.IsNull()) {
+            script = input.witness_utxo.scriptPubKey;
+        } else {
+            // There's no UTXO so we can just skip this now
```
Done

achow101 commented on src/wallet/psbtwallet.cpp:
```diff
@@ -39,12 +39,33 @@ TransactionError FillPSBT(const CWallet* pwallet, PartiallySignedTransaction& ps
+        input.FillSignatureData(sigdata);
+        const SigningProvider* provider = pwallet->GetSigningProvider(script, sigdata);
+        if (!provider) {
+            continue;
```
Done

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3740,11 +3766,11 @@ UniValue getaddressinfo(const JSONRPCRequest& request)
     isminetype mine = pwallet->IsMine(dest);
     ret.pushKV("ismine", bool(mine & ISMINE_SPENDABLE));
-    bool solvable = IsSolvable(*provider, scriptPubKey);
+    bool solvable = provider ? IsSolvable(*provider, scriptPubKey) : false;
```
In commit "Refactor: Require scriptPubKey to get wallet SigningProvider" (f7510c144b2262e1ee114eaa740b889c24ec7abb)

Would be slightly simpler to replace ternary with `solvable = provider && IsSolvable(*provider)`

achow101 replied:
Done

ryanofsky reviewed (approved):
Code review ACK f7510c144b2262e1ee114eaa740b889c24ec7abb. Pretty straightforward to verify this doesn't change behavior.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -550,7 +550,11 @@ static UniValue signmessage(const JSONRPCRequest& request)
     }
 
-    const SigningProvider* provider = pwallet->GetSigningProvider();
+    CScript script_pub_key = GetScriptForDestination(*pkhash);
```
I don't think it really matters as we still need `pkhash` since `signmessage` is supposed to only work on p2pkh addresses.

Sjors replied:
> we still need `pkhash` since `signmessage` is supposed to only work on `p2pkh` addresses

That will change in https://github.com/bitcoin/bitcoin/pull/16440/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR559

achow101 committed:
Accumulate result UniValue in SignTransaction

SignTransaction will be called multiple times in the future. Pass
it a result UniValue so that it can accumulate the results of multiple
SignTransaction passes.

ryanofsky reviewed (approved):
Code review ACK d0dab897afaac0a18aa47d3ce673a4a43a69178a. Thanks for the SignTransaction update. No other changes since last review

Sjors reviewed (approved):
Code review ACK d0dab897afaac0a18aa47d3ce673a4a43a69178a

ryanofsky commented:
@instagibbs and @meshcollider, you may be interested to review this. This is code you previously acked when it was part of #16341

promag reviewed (commented):
Code review ACK d0dab897afaac0a18aa47d3ce673a4a43a69178a.

instagibbs commented on src/interfaces/wallet.cpp:
```diff
@@ -117,8 +117,22 @@ class WalletImpl : public Wallet
     }
-    bool getPubKey(const CKeyID& address, CPubKey& pub_key) override { return m_wallet->GetLegacyScriptPubKeyMan()->GetPubKey(address, pub_key); }
-    bool getPrivKey(const CKeyID& address, CKey& key) override { return m_wallet->GetLegacyScriptPubKeyMan()->GetKey(address, key); }
+    bool getPubKey(const CScript& script, const CKeyID& address, CPubKey& pub_key) override
```
note for future work(TM): `address` is a horrible name for this argument

instagibbs commented on src/wallet/wallet.cpp:
```diff
@@ -2096,7 +2100,7 @@ void CWallet::AvailableCoins(interfaces::Chain::Lock& locked_chain, std::vector<
             }
 
-            const SigningProvider* provider = GetSigningProvider();
+            const SigningProvider* provider = GetSigningProvider(wtx.tx->vout[i].scriptPubKey);
```
do we need to check for knowledge of provider here?

instagibbs replied:
oh i cannot read.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3281,9 +3290,23 @@ UniValue signrawtransactionwithwallet(const JSONRPCRequest& request)
+            providers.insert(std::move(provider));
+        }
+    }
+    if (providers.size() == 0) {
```
non-blocking-nit: `.empty()`

achow101 replied:
Not fixing to avoid invalidating ACKs.

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/17371/commits/d0dab897afaac0a18aa47d3ce673a4a43a69178a

instagibbs commented:
ready for merge? @meshcollider

meshcollider commented on src/interfaces/wallet.cpp:
```diff
@@ -117,8 +117,22 @@ class WalletImpl : public Wallet
+    {
+        const SigningProvider* provider = m_wallet->GetSigningProvider(script);
+        if (provider) {
+            return provider->GetPubKey(address, pub_key);
```
Its kinda not that nice that there is no check the `script` and `address` have anything to do with each other. You could end up requesting the wrong SPKM for an address quite easily. Maybe move some of the ExtractDestination stuff inside here instead?

meshcollider commented on src/wallet/wallet.cpp:
```diff
@@ -1342,7 +1342,11 @@ bool CWallet::DummySignInput(CTxIn &tx_in, const CTxOut &txout, bool use_max_sig
-    const SigningProvider* provider = GetSigningProvider();
+    const SigningProvider* provider = GetSigningProvider(scriptPubKey);
+    if (!provider) {
+        // We don't know about this scriptpbuKey;
```
nit: typo (same below)

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2971,8 +2977,11 @@ static UniValue listunspent(const JSONRPCRequest& request)
         if (out.fSolvable) {
-            auto descriptor = InferDescriptor(scriptPubKey, *pwallet->GetLegacyScriptPubKeyMan());
-            entry.pushKV("desc", descriptor->ToString());
+            const SigningProvider* provider = pwallet->GetSigningProvider(scriptPubKey);
```
This can probably be combined with above to simplify

meshcollider reviewed (commented):
Code review ACK d0dab897afaac0a18aa47d3ce673a4a43a69178a
